### ISSUE TITLE
C++ bindings templates: Enums in map is broken

### ISSUE THREAD

**Issue by MitchRudominer** [domokit/mojo#468]
```
enum NormalEnum {
  FIRST,
  SECOND,
};

struct ContainsMapOfEnum {
  map<string, NormalEnum> map_of_enums;
};

```
yields the following C++ compilation error. 

gen/mojo/public/interfaces/bindings/tests/regression_tests.mojom.cc -o obj/out/Debug/gen/mojo/public/interfaces/bindings/tests/test_interfaces_cpp_sources.regression_tests.mojom.o
In file included from gen/mojo/public/interfaces/bindings/tests/regression_tests.mojom.cc:20:
../../mojo/public/cpp/bindings/lib/array_serialization.h:58:21: error: assigning to 'int' from incompatible type 'regression_tests::NormalEnum'
      output->at(i) = *it;
                    ^ ~~~
../../mojo/public/cpp/bindings/lib/map_serialization.h:161:57: note: in instantiation of function template specialization 'mojo::internal::ArraySerializer<regression_tests::NormalEnum, int, false>::SerializeElements<mojo::internal::MapValueIterator<mojo::String, regression_tests::NormalEnum>::Iterator>' requested here
        internal::ArraySerializer<MapValue, DataValue>::SerializeElements(
                                                        ^
gen/mojo/public/interfaces/bindings/tests/regression_tests.mojom.cc:2891:11: note: in instantiation of function template specialization 'mojo::SerializeMap_<mojo::String, regression_tests::NormalEnum, mojo::internal::Array_Data<char> *, int>' requested here
    mojo::SerializeMap_(

**Comment by MitchRudominer** [comment]
I won't have time to look at this today or tomorrow (promo committee stuff). If nobody picks it up in a few days I will look into it.

**Comment by vardhan-cr** [comment]
(cc'd johngro@)

mojom enums are now 'enum class's, so you can't serialize them as ints
without doing an explicit type conversion.  I think serializing enums into
Arrays is left unfixed.

**Comment by vardhan-cr** [comment]
A fix that works for me and is probably safe to check in (leaving it to
John's discretion though):

--- a/mojo/public/cpp/bindings/lib/array_serialization.h
+++ b/mojo/public/cpp/bindings/lib/array_serialization.h
@@ -55,7 +55,7 @@ struct ArraySerializer<E, F, false> {
     MOJO_DCHECK(!validate_params->element_validate_params)
         << "Primitive type should not have array validate params";
     for (size_t i = 0; i < num_elements; ++i, ++it)
-      output->at(i) = *it;
+      output->at(i) = static_cast<F>(*it);
   }

**Comment by johngro** [comment]
The binding generator currently ensures that the underlying data type of mojo enums is always an int32_t, so this is likely a perfectly fine fix.  I've assigned the bug to myself and will take a closer look in the morning, but something like this will probably be the fix.

I could use some help understanding how I missed this, however. I made sure that everything built and that I was passing all the tests before I checked in, so I'm not certain how this one slipped through the cracks.

**Comment by MitchRudominer** [comment]
Perhaps nobody is using this feature and we don't have a test for it.

**Comment by vardhan-cr** [comment]
^ what mitch said.
since Maps are not user-defined types (unlike structs and unions),
serialization code for it is not generated; it's in array_serialization.h.
 same goes for Arrays and Strings.

**Comment by johngro** [comment]
Indeed, that seems to be the case.  I found a good place to put a test for this; test_structs.mojom has a place where the types used for keys/values of maps/arrays are all exercises, but enums are not in the list of things tested.

I checked the mojom grammar, and according to it, enums should be allowed as both keys and values in these situations.  I went ahead and extended the test and immediately ran into a different issue.  Apparently the Dart binding generator really does not like it when you use an enum as either a key or a value for a map, or as a value for an array.  An example of it becoming upset with attempting to use an enum as a map's value type is given below.

I'm going to guess that Mitch is not seeing this because he is working on binding generator at the moment, but it does mean that I will not be able to check in tests unless I go ahead and fix the issues with the current generator.  I can probably work around the issues (generate just for c++) in order to exercise just this case and take a look at array_serialization.h, but it would be nice to be able to have an official test checked in.

What do you think?  Should I dive into the old dart binding generator, or just skip it since it will be replaced soon?

[1/41] ACTION //mojo/public/interfaces/bindings/tests:test_interfaces__generator(//build/toolchain/linux:clang_x64)
FAILED: python ../../mojo/public/tools/bindings/mojom_bindings_generator.py ../../mojo/public/interfaces/bindings/tests/test_structs.mojom --use_bundled_pylibs -d ../../ -I ../../ -I ../../ -o /usr/local/google/home/johngro/turquoise/mojo/src/out/Debug/gen
Traceback (most recent call last):
  File "../../mojo/public/tools/bindings/mojom_bindings_generator.py", line 221, in <module>
    sys.exit(main())
  File "../../mojo/public/tools/bindings/mojom_bindings_generator.py", line 215, in main
    processor.ProcessFile(args, remaining_args, generator_modules, filename)
  File "../../mojo/public/tools/bindings/mojom_bindings_generator.py", line 106, in ProcessFile
    filename)
  File "../../mojo/public/tools/bindings/mojom_bindings_generator.py", line 145, in _GenerateModule
    generator.GenerateFiles(filtered_args)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/mojom_dart_generator.py", line 446, in GenerateFiles
    lib_module = self.GenerateLibModule(args)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/pylib/mojom/generate/template_expander.py", line 56, in GeneratorInternal
    **kwargs)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/pylib/mojom/generate/template_expander.py", line 46, in ApplyTemplate
    return template.render(params)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/third_party/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/third_party/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/dart_templates/module.lib.tmpl", line 16, in top-level template code
    {%- include "module_definition.tmpl" %}
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/dart_templates/module_definition.tmpl", line 15, in top-level template code
    {{ struct_def(struct) }}
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/dart_templates/struct_definition.tmpl", line 66, in template
    {{encoding_macros.decode('result.' ~ packed_field.field|name, packed_field.field.kind, 8+packed_field.offset, packed_field.bit)|indent(6)}}
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/dart_templates/encoding_macros.tmpl", line 67, in template
    {{decode('values'~level, kind.value_kind|array('keys'~level~'.length'), 'bindings.ArrayDataHeader.kHeaderSize + bindings.kPointerSize', 0, level+1)|indent(4)}}
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/dart_templates/encoding_macros.tmpl", line 99, in template
    {{variable}} = decoder{{level}}.{{kind|decode_method(offset, bit)}};
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/mojom_dart_generator.py", line 306, in DecodeMethod
    methodName = _DecodeMethodName(kind)
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/mojom_dart_generator.py", line 300, in _DecodeMethodName
    return _DecodeMethodName(kind.kind) + 'Array'
  File "/usr/local/google/home/johngro/turquoise/mojo/src/mojo/public/tools/bindings/generators/mojom_dart_generator.py", line 305, in _DecodeMethodName
    return _spec_to_decode_method[kind.spec]
KeyError: 'x:mojo.test.MapValueTypes.EType'

**Comment by vardhan-cr** [comment]
That's odd -- compiling a mojom with "map<string, Enum>" works fine for me
(as in, the dart bindings are generated) after I patch the C++.  Maybe try
updating tree / gclient sync'ing?

**Comment by johngro** [comment]
that did it.  Looks like there was a fix from azani@ which went in yesterday afternoon that I was missing.  It was titled..

"""
Select the correct decode method for arrays of enums. #467
"""

Fix and tests for this issue are otw soon.

**Comment by johngro** [comment]
hey guys, sorry that took so long.

TL;DR
The fix is basically the cast which Vardhan recommended, also with some
type_traits trickery to make certain it is only applied in this specific
enum class case.  I also added some tests.

One thing to note...

When I tried to add any of the following to a .mojom file, it caused
problems with the JS and Dart unit tests.  I dodged having to fix this as
well by simply writing C++ unit tests which exercise the same code paths as
used by the C++ binding generators.

   - array<EnumType>
   - map<EnumKey, int>
   - map<int, EnumVal>

-john
